internal/cluster/cluster.go

```diff
@@ -777,10 +778,30 @@ func (m *Manager) DeleteClusterFiles(ctx context.Context, c *common.Cluster, obj
 	return nil
 }
 
-func (m Manager) PermanentClustersDeletion(ctx context.Context, olderThen strfmt.DateTime, objectHandler s3wrapper.API) error {
+func (m Manager) InactiveClusterDeregister(ctx context.Context, inactiveSince strfmt.DateTime) error {
+	log := logutil.FromContext(ctx, m.log)
+
+	var clusters []*common.Cluster
+	db := m.db.Unscoped()
+	if reply := db.Where("updated_at < ?", inactiveSince).Find(&clusters); reply.Error != nil {
```

Comment:
When does this field get updated? Is `updated_at` after agent events? user actions? 

In other words, when can a cluster become inactive?

Reply:
`updated_at` is a filed that managed by the DB ORM, every time that something in the cluster changes, validations, updates from the user, this filed will change.
If user create a cluster and don't do anything with or installation is done and all the logs have been collected nothing will happen in the cluster, this is how we know that it's not active.

Reply:
Events are not related because the are in a different DB table.

Reply:
Is there a risk for a cluster that is being used by a user to be considered inactive?

For example, say that I create a cluster and deploy an application on it and then don't do anything to the cluster. The application runs, everything works as expected, but no maintenance is needed. How do we know the cluster is active/inactive in this case? Logs collection from the cluster? Metrics?

Reply:
the logs and events are collected approximately every hour and the metrics are collected during the runtime. 
I guess that 30 days is enough time to collect whatever we need.

Reply:
@flaper87 i think that you are confusing between the result cluster and the cluster that we have. 
We don't delete users cluster, we just delete the assisted-installer representation of the cluster.
User applications are running on a cluster that we no longer have access to.

Reply:
> We don't delete users cluster, we just delete the assisted-installer representation of the cluster.
User applications are running on a cluster that we no longer have access to.

Yeah, I understood this. What I would like is to understand when `update_at` is actually updated to understand when a cluster is considered inactive (or unreachable). 

Using `updated_at` as a mechanism to monitor the cluster's heartbeat is a bit counterintuitive to me. Mostly because it doesn't have a documented, and dedicated logic meant to identify reachable/unreachable clusters.

Another question that may help clarify this (at least for me) is: 

Is the expectation that, at some point, we will always deregister/delete clusters from the AI DB? Or do we want to have them registered for the entire lifecycle of the cluster (and therefore be able to get heartbeats from the actual cluster)? 

Also, don't feel like we should hold-off this PR on my comments. I just want to learn more about the architecture, its expectations, and see if there's something else we may want to modify in follow-up PRs.

Reply:
We can look at our cluster definition as an installation attempt, after user finished the installation he no longer need the cluster, all he need to do is download kubeconfig and start working with it, there are no other operations that he can perform on it, except for day2 but it's done by creating a day2 cluster and we can look at it as another installation attempt. 

Because users are using our cluster only once we don't need to store it forever and because there are no operation that can be performed on it we don't need to archive it as well, so a simple solution will be to delete it. it will reduce our storage capacity and scale, because we perform  refresh status on installed draft and error clusters (log collection etc).

Regarding `updated_at` it is set automatically by gorm https://v1.gorm.io/docs/update.html, there are few fields that gorm can set automatically `crated_at` `deleted_at` `updated_at` we are just using it.
So every time something in the cluster is changing, any field in the cluster table, `updated_at` is changing as well, we are just using this mechanism.

Reply:
> We can look at our cluster definition as an installation attempt, after user finished the installation he no longer need the cluster, all he need to do is download kubeconfig and start working with it, there are no other operations that he can perform on it, except for day2 but it's done by creating a day2 cluster and we can look at it as another installation attempt.
>
> Because users are using our cluster only once we don't need to store it forever and because there are no operation that can be performed on it we don't need to archive it as well, so a simple solution will be to delete it. it will reduce our storage capacity and scale, because we perform refresh status on installed draft and error clusters (log collection etc).

This makes more sense, now. I was missing the part that we use the cluster once for deployment and then it's not needed anymore. Then for day2 operations, we have a separate cluster, which as you mentioned is a separate definition and therefore can be tracked separately. 

Is this documented somewhere? I would like to read more about it.

> Regarding updated_at it is set automatically by gorm https://v1.gorm.io/docs/update.html, there are few fields that gorm can set automatically crated_at deleted_at updated_at we are just using it.
So every time something in the cluster is changing, any field in the cluster table, updated_at is changing as well, we are just using this mechanism.

Yeah, I'm aware of this. This is a common pattern in most ORM frameworks. My question is not related to how the actual database update for `update_at` happens but rather what `updated_at` actually means. Using `updated_at` as a heartbeat pulse is a bit counterintuitive as the `updated_at` field is rarely able to hold the logic needed to consider a resource as "alive". It seems to work in this case, tho but I'm afraid this is just luck and not a deliberate design decision when the model/workflow was designed. Not something that needs to be acted on right now, but we may have to revisit this logic in the future.

Thank you for your comments and for sharing the details of the architecture:

![giphy](https://user-images.githubusercontent.com/13816/111276813-a84aee80-862f-11eb-81d3-c255e9235b4f.gif)

Reply:
> Is this documented somewhere? I would like to read more about it.
We have an HLD but i'm not sure if he is updated https://docs.google.com/document/d/1jxNMTlotmJ0GFZ1GUEQ3hfOOVLzMT6mvY8Ufo5RdErY/edit 

I agree that updated_at is an implementation details that we look at, but at this point we don't have any other indication.